File: day-12b/src/main.rs
use std::collections::HashSet;
use std::fmt::Display;
use std::fs::read_to_string;
use std::iter::repeat;
use std::str::FromStr;

use anyhow::{bail, Context, Ok, Result};
use cached::proc_macro::cached;
use itertools::Itertools;
use strum_macros::EnumIs;

#[derive(PartialEq, Eq, Hash, Clone, Copy, Debug, EnumIs)]
enum Condition {
    Damaged,
    Unknown,
    Operational,
}

impl Display for Condition {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let c = match self {
            Condition::Damaged => 'D',
            Condition::Operational => 'O',
            Condition::Unknown => 'U',
        };
        write!(f, "{c}")
    }
}

impl TryFrom<&char> for Condition {
    type Error = anyhow::Error;

    fn try_from(s: &char) -> Result<Self> {
        match s {
            '#' => Ok(Condition::Damaged),
            '?' => Ok(Condition::Unknown),
            '.' => Ok(Condition::Operational),
            _ => bail!("Can't construct a condition from {s}"),
        }
    }
}

#[cached]
fn num_possible_fits(contiguous_broken: Vec<u32>, conditions: Vec<Condition>) -> usize {
    if conditions.len() < contiguous_broken.len() {
        return 0;
    }

    if conditions[0].is_operational() {
        return num_possible_fits(contiguous_broken, conditions[1..].to_vec());
    }

    let grouped_by_operational: Vec<(bool, usize)> = conditions
        .iter()
        .group_by(|c| c.is_operational())
        .into_iter()
        .map(|(operational, group_iter)| (operational, group_iter.count()))
        .collect();
    debug_assert!(!grouped_by_operational[0].0);
    debug_assert!(!grouped_by_operational[grouped_by_operational.len() - 1].0);

    if (contiguous_broken.iter().sum::<u32>() as usize)
        > grouped_by_operational
            .iter()
            .filter(|(operational, _)| !operational)
            .map(|(_, group_length)| group_length)
            .sum()
    {
        return 0;
    }

    let grouped_by_condition: Vec<(&Condition, usize)> = conditions
        .iter()
        .group_by(|c| c.to_owned())
        .into_iter()
        .map(|(condition, group_iter)| (condition, group_iter.count()))
        .collect();
    debug_assert_ne!(grouped_by_condition[0].0, &Condition::Operational);
    debug_assert_ne!(
        grouped_by_condition[grouped_by_condition.len() - 1].0,
        &Condition::Operational
    );

    let first_contiguous = contiguous_broken[0] as usize;

    if grouped_by_operational[0].1 < first_contiguous {
        let first_operational_index = grouped_by_operational[0].1 + 1;
        if conditions[..first_operational_index].contains(&Condition::Damaged) {
            return 0;
        }
        return num_possible_fits(
            contiguous_broken,
            conditions[first_operational_index..].to_vec(),
        );
    }

    if grouped_by_operational[grouped_by_operational.len() - 1].1
        < (contiguous_broken[contiguous_broken.len() - 1] as usize)
    {
        let last_operational_index =
            conditions.len() - grouped_by_operational[grouped_by_operational.len() - 1].1 - 1;
        if conditions[last_operational_index..].contains(&Condition::Damaged) {
            return 0;
        }
        return num_possible_fits(
            contiguous_broken,
            conditions[..last_operational_index].to_vec(),
        );
    }

    let mut answer = 0;

    if contiguous_broken.len() == 1 {
        if grouped_by_condition.iter().any(|(c, _)| c.is_damaged()) {
            for i in 0..conditions.len() {
                if i != 0 && conditions[i - 1].is_damaged() {
                    break;
                }

                if let Some(slice) = conditions.get((i + first_contiguous)..) {
                    if slice.contains(&Condition::Damaged) {
                        continue;
                    }
                }

                match conditions.get(i..(i + first_contiguous)) {
                    Some(slice) => {
                        if slice.len() < first_contiguous {
                            break;
                        }
                        let to_test: HashSet<&Condition> = HashSet::from_iter(slice);
                        if to_test.contains(&Condition::Operational) {
                            continue;
                        }
                        if !to_test.contains(&Condition::Damaged) {
                            continue;
                        }
                    }
                    None => break,
                }

                answer += 1
            }
        } else {
            for (condition, group_length) in grouped_by_condition {
                if condition.is_unknown() && group_length >= first_contiguous {
                    answer += (group_length - first_contiguous) + 1
                }
            }
        }
    } else {
        let range_to_test = grouped_by_operational[0].1 - first_contiguous + 1;
        for i in 0..range_to_test {
            if i != 0 && conditions[i - 1].is_damaged() {
                break;
            }
            if let Some(Condition::Damaged) = conditions.get(i + first_contiguous) {
                continue;
            }
            if let Some(slice) = conditions.get((i + first_contiguous + 1)..) {
                answer += num_possible_fits(contiguous_broken[1..].to_vec(), slice.to_vec())
            }
        }

        if conditions[..range_to_test].iter().all(|c| c.is_unknown()) {
            answer += num_possible_fits(contiguous_broken, conditions[range_to_test..].to_vec())
        }
    }
    answer
}

fn find_conditions(string: &str) -> Result<Vec<Condition>> {
    let re = regex::Regex::new(r"\.+").unwrap();
    let modded_string = re.replace_all(string, ".");
    modded_string
        .trim_matches('.')
        .chars()
        .map(|c| Condition::try_from(&c))
        .collect()
}

#[derive(Clone)]
struct Row {
    conditions: Vec<Condition>,
    contiguous_broken_groups: Vec<u32>,
}

impl Row {
    fn num_possible_arrangements(self) -> usize {
        num_possible_fits(self.contiguous_broken_groups, self.conditions)
    }
}

const REPEATS: usize = 5;

impl FromStr for Row {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self> {
        let (left, right) = match s.split(' ').collect_vec()[..] {
            [left, right] => (left, right),
            _ => bail!("Couldn't parse {s} into a row"),
        };
        let conditions = find_conditions(repeat(left).take(REPEATS).join("?").as_str())?;
        let contiguous_broken_groups = repeat(right)
            .take(REPEATS)
            .join(",")
            .split(',')
            .map(|val| val.parse())
            .collect::<Result<_, _>>()?;
        Ok(Row {
            conditions,
            contiguous_broken_groups,
        })
    }
}

fn read_input(filename: &str) -> Result<String> {
    read_to_string(filename).with_context(|| format!("Expected {filename} to exist!"))
}

fn solve(filename: &str) -> usize {
    read_input(filename)
        .unwrap()
        .lines()
        .map(|line| Row::from_str(line).unwrap().num_possible_arrangements())
        .sum()
}

fn main() {
    println!("{}", solve("input.txt"))
}

File: day-12b/src/main.py
import re
from collections.abc import Iterator
from contextlib import suppress
from dataclasses import dataclass
from enum import Enum
from functools import cache
from itertools import groupby, repeat
from typing import Final, Self

import pytest


class Condition(Enum):
    DAMAGED = "#"
    UNKNOWN = "?"
    OPERATIONAL = "."

    @cache
    def is_operational(self) -> bool:
        return self is Condition.OPERATIONAL

    def __repr__(self) -> str:
        return self.name[0]


@cache
def num_possible_fits(
    contiguous_broken: tuple[int, ...], conditions: tuple[Condition, ...]
) -> int:
    if len(conditions) < len(contiguous_broken):
        return 0

    if conditions[0] is Condition.OPERATIONAL:
        return num_possible_fits(contiguous_broken, conditions[1:])

    grouped_by_operational = [
        (operational, sum(1 for _ in group_iter))
        for operational, group_iter in groupby(conditions, key=Condition.is_operational)
    ]

    if sum(contiguous_broken) > sum(
        group_length
        for operational, group_length in grouped_by_operational
        if not operational
    ):
        return 0

    grouped_by_condition = [
        (condition, sum(1 for _ in group_iter))
        for condition, group_iter in groupby(conditions)
    ]

    first_contiguous = contiguous_broken[0]

    # If the first group of (Damaged + Unknown) doesn't fit the first contiguous group:
    if grouped_by_operational[0][1] < first_contiguous:
        first_operational_index = grouped_by_operational[0][1] + 1
        if Condition.DAMAGED in conditions[:first_operational_index]:
            return 0
        return num_possible_fits(
            contiguous_broken, conditions[first_operational_index:]
        )

    # If the last group of (Damaged + Unknown) doesn't fit the last contiguous group:
    if grouped_by_operational[-1][1] < contiguous_broken[-1]:
        last_operational_index = len(conditions) - grouped_by_operational[-1][1] - 1
        if Condition.DAMAGED in conditions[last_operational_index:]:
            return 0
        return num_possible_fits(contiguous_broken, conditions[:last_operational_index])

    answer = 0

    if len(contiguous_broken) == 1:
        if any(condition is Condition.DAMAGED for condition, _ in grouped_by_condition):
            for i in range(len(conditions)):
                if i != 0 and conditions[i - 1] is Condition.DAMAGED:
                    break
                if Condition.DAMAGED in conditions[(i + first_contiguous) :]:
                    continue
                slice_to_test = conditions[i : (i + first_contiguous)]
                if len(slice_to_test) < first_contiguous:
                    break
                to_test = set(slice_to_test)
                if Condition.OPERATIONAL in to_test:
                    continue
                if Condition.DAMAGED not in to_test:
                    continue
                answer += 1
        else:
            for condition, group_length in grouped_by_condition:
                if (
                    condition is Condition.UNKNOWN
                    and group_length >= first_contiguous
                ):
                    answer += (group_length - first_contiguous) + 1
    else:
        # We now have a sequence of >1 contiguous groups, where:
        # (1) the first group of (Damaged + Unknown)
        #     is at least big enough to fit the first contiguous group, and;
        # (2) the last group of (Damaged + Unknown)
        #     is at least big enough to fit the last contiguous group
        range_to_test = (grouped_by_operational[0][1] - first_contiguous) + 1
        for i in range(range_to_test):
            if i != 0 and conditions[i - 1] is Condition.DAMAGED:
                break
            with suppress(IndexError):
                if conditions[i + first_contiguous] is Condition.DAMAGED:
                    continue
            remaining = conditions[(i + first_contiguous + 1) :]
            answer += num_possible_fits(contiguous_broken[1:], remaining)

        if set(conditions[:range_to_test]) == {Condition.UNKNOWN}:
            answer += num_possible_fits(contiguous_broken, conditions[range_to_test:])

    return answer


REPEATS: Final = 5


def find_conditions(string: str) -> Iterator[Condition]:
    for char in re.sub(r"\.+", ".", string).strip("."):
        yield Condition(char)


@dataclass(slots=True, kw_only=True, frozen=True)
class Row:
    conditions: tuple[Condition, ...]
    contiguous_broken_groups: tuple[int, ...]

    @classmethod
    def from_string(cls, line: str) -> Self:
        left, _, right = line.partition(" ")
        conditions = tuple(find_conditions("?".join(repeat(left, REPEATS))))
        contiguous_broken_groups = tuple(
            int(val) for val in ",".join(repeat(right, REPEATS)).split(",")
        )
        return cls(
            conditions=conditions, contiguous_broken_groups=contiguous_broken_groups
        )

    def num_possible_arrangements(self) -> int:
        return num_possible_fits(self.contiguous_broken_groups, self.conditions)


def parse_input(filename: str) -> Iterator[Row]:
    with open(filename) as f:
        for line in f:
            yield Row.from_string(line)


def solve(filename: str) -> int:
    return sum(row.num_possible_arrangements() for row in parse_input(filename))


def main() -> None:
    print(solve("input.txt"))


if __name__ == "__main__":
    main()


@pytest.mark.parametrize(
    ("input", "solution"),
    [
        ("???.### 1,1,3", 1),
        (".??..??...?##. 1,1,3", 16384),
        ("?#?#?#?#?#?#?#? 1,3,1,6", 1),
        ("????.#...#... 4,1,1", 16),
        ("????.######..#####. 1,6,5", 2500),
        ("?###???????? 3,2,1", 506250),
    ],
)
def test_examples(input: str, solution: int) -> None:
    row = Row.from_string(input)
    assert row.num_possible_arrangements() == solution
